Various fixes

poire-z opened the pull request:
As discussed in some issues some days ago:
- `disable-wifi.sh `: reduced sleep time (1 second was noticable when exiting koreader) : 2x200ms is enough for my device to not crash, and should be less noticable to those who dont need it.
- clear reader's cache: also clear in-memory cache (I think it's quite simple, and less risky than exiting reader without flushing settings). I thought about adding `Cache.cached={} `in its `Cache:clear() `method, but this method is used by unit tests that expect `cached `to not be reset between calls to `clear()`. So, yes, i access some unexposed object attribute from outside, but we're in "developer options" :) If you prefer I add an other method for that, let me know.

About textboxwidget merged yesterday: I wasn't proud of my unit test results, and I thought i may also pass previous char to isSplitable, and it indeed gives better wrapping.

houqp commented on frontend/ui/widget/textboxwidget.lua:
```diff
@@ -119,9 +119,10 @@ function TextBoxWidget:_splitCharWidthList()
-            if util.isSplitable(c, next_c) then
+            -- We give next and prev chars to isSplitable() for a wiser decision
+            local next_c = idx+1 <= size and self.char_width_list[idx+1].char or false
+            local prev_c = idx-1 >= 1 and self.char_width_list[idx-1].char or false
```
Instead of lookup prev_c from the table every time, can we remember the previous char in a local variable instead?

houqp commented on frontend/apps/filemanager/filemanagermenu.lua:
```diff
@@ -213,6 +213,13 @@ function FileManagerMenu:setUpdateItemTable()
+                            -- Clear Cache object, so it doesn't serialize
+                            -- anything old on exit
+                            local Cache = require("cache")
+                            Cache:clear()
```
I think this is not going to be safe because `clear` calls `onFree` method on all the cached objects. Some of those objects are allocated through malloc, which means those references will have a dangling pointer to a freed memory. It's really hard to do this right without restarting the reader from scratch. I still think we should create a prompt for confirmation before the purge and force user to restart the reader after the purge.

poire-z replied:
There's already this prompt `text = _("Clear cache/ and cr3cache/ ?")`.
Should I add "Please restart koreader now." to this message or in an InfoMessage after it's done ?
Or do you have another idea for "forcing user to restart" ? :)

Also, to avoid this crash on exit (sorting according to file timestamps for files in cached but no more on disk, so with lfs.attributes("access") = nil):
```
./luajit: frontend/cache.lua:159: attempt to compare two nil values
stack traceback:
        frontend/cache.lua:159: in function <frontend/cache.lua:159>
        [C]: in function 'sort'
        frontend/cache.lua:159: in function 'serialize'
```
I think it's still needed to just do `Cache.cached = {}` .
Would that be fine ?

houqp replied:
> There's already this prompt text = _("Clear cache/ and cr3cache/ ?").
> Should I add "Please restart koreader now." to this message or in an InfoMessage after it's done ?
> Or do you have another idea for "forcing user to restart" ? :)

Showing please restart KOReader should be good enough for now. We don't have a function to restart koreader automatically for the user unfortunately :(

> I think it's still needed to just do Cache.cached = {} .

Yes, let's keep this one :)

houqp commented:
Nice catch, I have been wondering why the new build exits slower than the previous ones ;)

Also thanks for writing more test cases. Do you have the emulator environment setup now?

poire-z commented:
> Do you have the emulator environment setup now?

No... :| (low on time and i learned enough new stuff these last weeks :) well...)
For unit tests, i just pasted some code in readerui.lua and print the result, that I put back if fine in the unit test...

houqp commented:
Haha, that's cool. Just curious how you verify the unit test locally. That's one way to do it :P

poire-z commented:
Changes since previous commits:
- disable-wifi.sh: checks if kernel module is really loaded before usleep and rmmod, so exiting koreader when wifi was off shouldn't pause at all.
- clear readers's cache: did as suggested.

textboxwidget:
I ended up rewritting the logic flow : each time i got to that part of the code, i spent a few minutes rereading to understand how it works - i could have been ok with just adding some comments, but well.
I changed it from:` if then A else (repeat until, if then A else B)` end
to a simpler: `while end if then A else B`, with miminam table lookups.
But the flow and results should be the same (if i didnt mess up). Dict and Wikipedia results looks as fine as before.
If you prefer the previous version, i'll have no problem puting it back :)

Also, I found out CJK text also have some rules for wrapping, with chars not allowed at start or end of line, really similar to what I did for non CJK.
https://en.wikipedia.org/wiki/Line_breaking_rules_in_East_Asian_languages
I implemented it in this [util.lua.txt](https://github.com/koreader/koreader/files/650237/util.lua.txt) but it is not part of this PR (the CJK unit test would probably need to be updated - i'd rather let that to CJK native readers :)
It looks fine to me on zh/ko/ja wikipedia texts, but I have absolutely no idea if it's really better, if the rules are good, and if my cut and paste of these chars were done well.

houqp commented:
Rewriting bad code into more maintainable version is always welcome! I will give a closer look at the patch after work. Also, nice find on the CJK line wrapping rules. I will look into that too.

houqp commented:
Thanks! That part of the logic looks much cleaner now ;)

houqp commented:
BTW, your CJK patches looks good to me, feel free to send a PR. I can help fix the unit tests. Today I learned that some of the characters in Japanese are not splitable :P
